litentry/core/ruleset-build/src/lib.rs

```diff
@@ -1,8 +1,78 @@
-#[cfg(test)]
-mod tests {
-	#[test]
-	fn it_works() {
-		let result = 2 + 2;
-		assert_eq!(result, 4);
+// Copyright 2020-2022 Litentry Technologies GmbH.
+// This file is part of Litentry.
+//
+// Litentry is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(all(feature = "std", feature = "sgx"))]
+compile_error!("feature \"std\" and feature \"sgx\" cannot be enabled at the same time");
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+extern crate sgx_tstd as std;
+
+// re-export module to properly feature gate sgx and regular std environment
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+pub mod sgx_reexport_prelude {
+	pub use thiserror_sgx as thiserror;
+	pub use url_sgx as url;
+}
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::*;
+
+use itc_rest_client::{error::Error as HttpError, RestPath};
+use serde::{Deserialize, Serialize};
+use std::{fmt::Debug, str, string::String};
+
+use itp_types::AccountId;
+
+use litentry_primitives::{Identity, Ruleset};
+
+pub mod discord;
+
+use crate::discord::ruleset2_verification;
+
+pub type Result<T> = core::result::Result<T, Error>;
+
+#[derive(Debug, thiserror::Error, Clone)]
+pub enum Error {
+	#[error("Ruleset error: {0}")]
+	Ruleset1Error(String),
+
+	#[error("Other error: {0}")]
+	RulesetOtherError(String),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct CheckJoinDiscordResponse {
+	data: bool,
+	message: String,
+	has_errors: bool,
+	msg_code: u32,
+	success: bool,
+}
+
+impl RestPath<String> for CheckJoinDiscordResponse {
+	fn get_path(path: String) -> core::result::Result<String, HttpError> {
+		Ok(path)
+	}
+}
+
+pub fn ruleset_verify(who: AccountId, identity: Identity, ruleset: Ruleset) -> Result<()> {
```

Comment:
We have two problems here:
- `identity` is some single identity. What if we need to calculate something over all relevant identities? E.g. ruleset 10: WBTCs over all evm accounts. So you need to always pass `vec_identity` here, or what's better: each ruleset should have its own way to retrieve selected identities, which means `vec_identity` should be always the full linked identities for **generality**(instead of discord only -- since to "select" out discord identities, you already rely on the ruleset::R2 to do it, why not process it in one place). But then you don't have pass in `vec_identity` at all as it can always be retrieved by `who`
- we are missing the parameter (placeholder), notice this is the entrance for ruleset computation

Reply:
For `identity`, agree. 
`Parameter (placeholder)` is included inside enum `Ruleset`.